src/jsn.erl

```diff
@@ -656,6 +686,38 @@ from_proplist(X, _Options) -> X.
 %% internal functions
 %%==============================================================================
 
+-spec apply_condition(condition()|conditions(), json_term()) -> boolean().
+apply_condition({Path, ConditionFun}, Element) when is_function(ConditionFun, 1) ->
+    ConditionFun(get(Path, Element));
+apply_condition({Path, Value}, Element) ->
+    get(Path, Element) == Value;
+apply_condition(ConditionFun, Element) when is_function(ConditionFun, 1) ->
+    ConditionFun(Element);
+apply_condition([Condition|Rest], Element) ->
+    case apply_condition(Condition, Element) of
+        true ->
+            apply_condition(Rest, Element);
+        false ->
+            false
+    end;
+apply_condition([], _Element) ->
+    true;
+apply_condition(_, _) ->
+    erlang:error(invalid_condition).
```

Comment:
So, I am of two minds on this. On the one hand, as is, it might as well be a `badarg`:

```erlang
> try jsn:select([{value, <<"username">>}], [fuuu], [jsn:new()]) catch C:E -> T = erlang:get_stacktrace(), io:format("Exception ~p:~p, trace: ~p~n", [C,E,T]) end.
Exception error:invalid_condition, trace: [{jsn,apply_condition,2,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,706}]},
                                           {jsn,apply_condition,2,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,697}]},
                                           {jsn,'-select/3-fun-0-',3,
                                            [{file,
                                              "/Users/nal/src/personal/erlang/jsn/_build/default/lib/jsn/src/jsn.erl"},
                                             {line,195}]},
                                           {lists,filtermap,2,
                                            [{file,"lists.erl"},{line,1317}]},
                                           {erl_eval,do_apply,6,
                                            [{file,"erl_eval.erl"},
                                             {line,674}]},
                                           {erl_eval,try_clauses,8,
                                            [{file,"erl_eval.erl"},
                                             {line,904}]},
                                           {shell,exprs,7,
                                            [{file,"shell.erl"},{line,686}]},
                                           {shell,eval_exprs,7,
                                            [{file,"shell.erl"},{line,641}]}]
ok
```

I guess the `invalid_condition` makes it obvious that the problem is a condition, but not which one. I feel like it should be either a bare `badarg` exception (the trace will make it obvious where it is), or, if we use a custom error, it should include the broken arg, i.e.

apply_condition(C, _) ->
    erlang:error({invalid_condition, C}).

Reply:
@mrallen1 curious on your thoughts here.

Reply:
I would prefer the error report the precise `invalid_condition`. The stack trace is nice and all, but precision is better, imo.